### PR TITLE
[Hub Generated] Review request for Microsoft.IoTCentral to add version 2018-09-01

### PR DESCRIPTION
**Pull request by kawayiboy**
If you are a MSFT employee you can view your [work branch via this link](https://portal.azure-devex-tools.com/app/branch/kawayiboy/azure-rest-api-specs/telu-templatesapi).

### Contribution checklist:
- [ ] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (kawayiboy) [inline specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: Still discussing about making it regional or global

**Comment by adxsdknet** [comment]
# Automation for azure-sdk-for-net 
A PR has been created for you: 
https://github.com/azure/azure-sdk-for-net/pull/5615 
.NET SDK Commits: 
https://github.com/adxsdknet/azure-sdk-for-net/pull/5615/commits/dc9a35d69b4cce24b585834a08159fc6f70a2191

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (sergey-shandar) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: let me know when you have a decision.

Comment (majastrz) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: Right now, it's modeled as a top-level subscription resource. Such resources have to be ProxyOnly. ARM doesn't provide regional routing for top-level ProxyOnly resources, which will force them to be global.

If this is undecided, should we hold off on reviewing this PR?

---
In reply to: [270201167](https://github.com/Azure/azure-rest-api-specs/pull/5495#discussion_r270201167) [](ancestors = 270201167)

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (majastrz) [inline specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: >filtered [](start = 61, length = 8)

What does the word "filtered" refer to here? What filters the templates?

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (majastrz) [inline specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: >AppTemplate [](start = 5, length = 11)

The shape of this resource doesn't actually confirm to the RPC. Take a look at [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#get-resource](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#get-resource). 

Your resource should look like this example:
```
{
  "name": "{templateName}",
  "type": "Microsoft.IoTCentral/applicationTemplates",
  "id": "/subscriptions/{subscriptionId}/providers/Microsoft.IoTCentral/applicationTemplates/{templateName},
  "properties": {
     <this is where your custom properties go>
  }
}
```

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (majastrz) [inline specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: >manifestId [](start = 9, length = 10)

Is this supposed to be the identifier of the resource that's present in the GET route? If yes, then that's not allowed. (See my other comment.)

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (majastrz) [inline specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: >title [](start = 9, length = 5)

What is the difference between `name` and `title`?

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (majastrz) [inline specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: >name [](start = 9, length = 4)

`name` is generally considered a reserved property name and intended to be a top level property (see the comment with the json example). You should call it `appTemplateName` instead to avoid user confusion. Microsoft.Authorization RP follows a similar pattern.

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (majastrz) [inline specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: >application templates [](start = 42, length = 21)

Is "application template" an official name for this IOT central feature? Could a user confuse "application template" with ARM templates easily?

**Review by majastrz — CHANGES_REQUESTED** [review]
The shape of the resource violates the RPC. I added comments with details. Can you take a look?

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (kawayiboy) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: title is more like displayName showing on the UI

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (kawayiboy) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: it's for our internal references, I removed it

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (kawayiboy) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: changed it to IoTAppTemplates

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (kawayiboy) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: it's not the identifier of the resource, and I changed the structure accordingly

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (kawayiboy) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: I've changed "name" based on your suggestion

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (kawayiboy) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: And we can not onboard get single template resource api now, we haven't implemented yet.

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (kawayiboy) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: @sergey-shandar @majastrz , we have decided to go with this global format

Comment (majastrz) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: Makes sense.

---
In reply to: [271399838](https://github.com/Azure/azure-rest-api-specs/pull/5495#discussion_r271399838) [](ancestors = 271399838)

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (majastrz) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: Sorry for the confusion. I was just asking the question rather than making a suggestion to change the naming. "IoTAppTemplates" actually violates the naming rules because we require camelCasing. I'm ok with either "appTemplates" or "applicationTemplates".

---
In reply to: [271079702](https://github.com/Azure/azure-rest-api-specs/pull/5495#discussion_r271079702) [](ancestors = 271079702)

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (majastrz) [inline specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: You should model all the properties so the generated SDK can take advantage of strong typing for languages that support it. In the example you have properties "manifestId", "manifestVersion", "appTemplateName", "title", "order", and "description". You should create a separate type called `AppTemplateProperties` and reference it here.

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (majastrz) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: Yes, you can. Once we approve this Swagger, you should hold off on asking for it to be merged. You just need signed off swagger PR to unblock your RP manifest change. Once you implement the service change to GET single template, you can ask to have the swagger PR merged to master.

---
In reply to: [271089778](https://github.com/Azure/azure-rest-api-specs/pull/5495#discussion_r271089778) [](ancestors = 271089778)

**Review by majastrz — CHANGES_REQUESTED** [review]
I left more comments to be addressed.
@majastrz ,I resolved those, could you take a look again, thanks

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (kawayiboy) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: So would this name "IoTAppTemplates" work? we discuss it out

Comment (majastrz) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: Yeah iotAppTemplates would be ok too. (ARM resource IDs are expected to be case-insensitive, so your service should accept all variations.)

**Review by KrisBash — APPROVED** [review]
Marcin's comments appear to be addressed

**Review by KrisBash — CHANGES_REQUESTED** [review]
Discussing with Marcin, there may be more changes needed

**Comment by majastrz** [comment]
@kawayiboy, I was talking to @KrisBash about the single GET requirement. Here's a couple of questions:
1. Are iotAppTemplates always read only? How do they get created?
2. Do you ever need RBAC to filter out individual templates per user's permissions or is it ok if all of them get returned if the user has /read on the subscription?

**Comment by kawayiboy** [comment]
> @kawayiboy, I was talking to @KrisBash about the single GET requirement. Here's a couple of questions:
> 
> 1. Are iotAppTemplates always read only? How do they get created?
> 2. Do you ever need RBAC to filter out individual templates per user's permissions or is it ok if all of them get returned if the user has /read on the subscription?

Hi @majastrz , iotAppTemplates are readonly ,and they are created and managed by us(our team)
And all of templates get returned if the user has /read on the subscription

**Comment by majastrz** [comment]
> > @kawayiboy, I was talking to @KrisBash about the single GET requirement. Here's a couple of questions:
> > 
> > 1. Are iotAppTemplates always read only? How do they get created?
> > 2. Do you ever need RBAC to filter out individual templates per user's permissions or is it ok if all of them get returned if the user has /read on the subscription?
> 
> Hi @majastrz , iotAppTemplates are readonly ,and they are created and managed by us(our team)
> And all of templates get returned if the user has /read on the subscription

Thanks for confirming. I discussed this yesterday with @KrisBash and we are going to officially formalize a pattern for APIs like yours to reduce how much needs to be implemented there. 

I didn't see the changes we discussed over IM last night, so here's a summary:
- Forget about the GET instance route. GET collection is enough for this type of API
- Remove the "id", "name", and "type" properties because they don't make sense anymore since iotAppTemplates is not a resource.
- Change the payload to get rid of "properties" and move everything up to top level. This is now ok because we're not treating it as a resource anymore. With this pattern, the payload of individual iotAppTemplate can be anythng that makes sense for you as long as you do not include any reserved ARM properties like "name", "id", etc.
- The collection response should contain an array property called "value" because paging may be needed in the future.

Once these changes are in, both @KrisBash and I will approve the PR and unblock you. I will be checking email over the weekend and would be able to approve if I see a GitHub notification about changes being pushed or a new comment.

**Review by majastrz — CHANGES_REQUESTED** [review]
Left the comment based on our IM discussion last night. Take a look.

**Review by majastrz — APPROVED** [review]
Looks good.

**Comment by majastrz** [comment]
@KrisBash, this looks good now. I signed off. Can you also sign off?

**Comment by kawayiboy** [comment]
> Discussing with Marcin, there may be more changes needed

Done the changes

**Comment by kawayiboy** [comment]
We gonna test it on the dogfood, please hold off from merging this PR, thanks

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (kawayiboy) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: As we have already have Microsoft.IoTCentral as the namespace, we decide to keep the url as appTemplates. Thanks

**Comment by kawayiboy** [comment]
@KrisBash @majastrz , I have successfully test the api through ARMClient and Ibiza, could you sign off on this Swagger PR and help us merge it? Thanks
armclient calling,
armclient POST subscriptions/34d077e3-2c1a-48f5-8629-ffc0e657e0ed/providers/Microsoft.IoTCentral/appTemplates?api-version=2018-09-01
{
  "value": [
    {
      "manifestId": "iotc-default",
      "manifestVersion": "1.0.0",
      "name": "IoT Central Default template",
      "title": "Custom Application",
      "order": 3,
      "description": "Start with a blank template and define your application from scratch."
    },
    {
      "manifestId": "iotc-demo",
      "manifestVersion": "1.0.0",
      "name": "IoT Central Contoso Demo template",
      "title": "Sample Contoso",
      "order": 1,
      "description": "Get started with a predefined application for a connected device."
    },
    {
      "manifestId": "iotc-devkit-sample",
      "manifestVersion": "1.0.0",
      "name": "IoT Central Sample Developer Kit template",
      "title": "Sample Devkits",
      "order": 2,
      "description": "Want to connect a Raspberry PI or MXChip IoT DevKit? Start with this predefined app and get them connected in minutes."
    }
  ],
  "nextLink": null
}
Ibiza calling,
https://msazure.visualstudio.com/b32aa71e-8ed2-41b2-9d77-5bc261222004/_apis/git/repositories/8738e34d-71d0-42ce-9589-db63b5a5fe58/pullRequests/1522021/attachments/Ibiza-templatesapi2.gif

File: specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json
Comment (kawayiboy) [reply specification/iotcentral/resource-manager/Microsoft.IoTCentral/stable/2018-09-01/iotcentral.json]: > let me know when you have a decision. #Closed

we just go with this format

**Comment by kawayiboy** [comment]
@sergey-shandar, Marcin has signed off on this, could you help us merge it in? Thank you

**Comment by sergey-shandar** [comment]
@kawayiboy why it's assigned to @majastrz ?

**Comment by sergey-shandar** [comment]
@KrisBash could you review the changes?

**Comment by sergey-shandar** [comment]
@kawayiboy every PR should be assigned to Azure SDK swagger reviews team. Please, don't manually reassign.

**Review by sergey-shandar — APPROVED** [review]

**Review by KrisBash — APPROVED** [review]
LGTM

**Pull request merged**
